convert/convert.go

```diff
@@ -49,80 +49,105 @@ func File(file *hcl.File, options Options) ([]byte, error) {
-	// check if the next character is an ending paren, and include it if it is.
-	end := r.End.Byte
-	if c.bytes[end] == ')' {
-		end++
+func convertFile(file *hcl.File, options Options) (jsonObj, error) {
+	body, ok := file.Body.(*hclsyntax.Body)
+	if !ok {
+		return nil, fmt.Errorf("convert file body to body type")
 	}
-	return string(c.bytes[r.Start.Byte:end])
+
+	c := converter{
+		bytes:   file.Bytes,
+		options: options,
+	}
+
+	out, err := c.convertBody(body)
+	if err != nil {
+		return nil, fmt.Errorf("convert body: %w", err)
+	}
+
+	return out, nil
 }
 
 func (c *converter) convertBody(body *hclsyntax.Body) (jsonObj, error) {
-	var err error
 	out := make(jsonObj)
-	for key, value := range body.Attributes {
-		out[key], err = c.convertExpression(value.Expr)
-		if err != nil {
-			return nil, err
+
+	for _, block := range body.Blocks {
+		if err := c.convertBlock(block, out); err != nil {
+			return nil, fmt.Errorf("convert block: %w", err)
 		}
 	}
 
-	for _, block := range body.Blocks {
-		err = c.convertBlock(block, out)
+	var err error
+	for key, value := range body.Attributes {
+		out[key], err = c.convertExpression(value.Expr)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("convert expression: %w", err)
 		}
 	}
 
 	return out, nil
 }
 
-func (c *converter) convertBlock(block *hclsyntax.Block, out jsonObj) error {
-	var key string = block.Type
-
-	value, err := c.convertBody(block.Body)
-	if err != nil {
-		return err
+func (c *converter) rangeSource(r hcl.Range) string {
+	// for some reason the range doesn't include the ending paren, so
+	// check if the next character is an ending paren, and include it if it is.
+	end := r.End.Byte
+	if c.bytes[end] == ')' {
+		end++
 	}
+	return string(c.bytes[r.Start.Byte:end])
+}
 
+func (c *converter) convertBlock(block *hclsyntax.Block, out jsonObj) error {
+	key := block.Type
 	for _, label := range block.Labels {
-		if inner, exists := out[key]; exists {
+
+		// Labels represented in HCL are defined as quoted strings after the name of the block:
+		// block "label_one" "label_two"
+		//
+		// Labels represtend in JSON are nested one after the other:
+		// "label_one": {
+		//   "label_two": {}
+		// }
+		//
+		// To create the JSON representation, check to see if the label exists in the current output:
+		//
+		// When the label exists, move onto the next label reference.
+		// When a label does not exist, create the label in the output and set that as the next label reference
+		// in order to append (potential) labels to it.
+		if _, exists := out[key]; exists {
 			var ok bool
-			out, ok = inner.(jsonObj)
+			out, ok = out[key].(jsonObj)
 			if !ok {
-				// TODO: better diagnostics
 				return fmt.Errorf("Unable to convert Block to JSON: %v.%v", block.Type, strings.Join(block.Labels, "."))
 			}
 		} else {
-			obj := make(jsonObj)
-			out[key] = obj
-			out = obj
+			out[key] = make(jsonObj)
+			out = out[key].(jsonObj)
```

Comment:
again, why change?

Reply:
For both of these, this tripped me up for quite a bit. 

When using `out[key] =`, we're actually changing the value passed in by the user.
When using `out =`, we're just setting the "next" out value to get the next label.

All of this makes sense in hindsight, but creating the single `obj` and setting it to both `out[key]` as well as `out` wasn't immediately clear. 

This feels more natural in that we're creating a new value at `out[key]` and then setting `out` to that value that we _just created_.

I'm more than happy to revert any of this. It really was just trying to organize things in my head, but I don't want to make it confusing for you if the old way made more sense to you.